### PR TITLE
Completer setup guide

### LINKED ISSUE #120: Setup guide for system admins
Things a system admin should do before users can start using the ewaterycycle package should be documented.

1. Create conda env with ewatercycle package
2. install Docker / singularity
3. Pull /build container images #119
4. Download example parameter sets
5. Download example forcing for Marrmot
5. Prepare other parameter sets
4. Setup ESMValTool
5. Download OBS ERA5 with era5cli
3. Create /etc/ewatrercycle.yaml

### PR DESCRIPTION
**Pull request by sverhoeven**
Fixes #120

docs preview: https://ewatercycle--150.org.readthedocs.build/en/150/system_setup.html

### CODE REVIEW

**Comment by Peter9192** [comment]
Perhaps also move the setup guide one level up in the RTD toctree? And should it somehow replace (part of) https://ewatercycle.readthedocs.io/en/latest/hpc_to_cluster.html?

**Comment by sverhoeven** [comment]
> Perhaps also move the setup guide one level up in the RTD toctree? And should it somehow replace (part of) https://ewatercycle.readthedocs.io/en/latest/hpc_to_cluster.html?

It is now the first chapter. I also switched it to RST because all those instructions dont really make sense in a notebook. Only overlap with hpc_to_cluster is conda installation which is now linked.

File: docs/system_setup.rst [hunk 1]
```diff
@@ -0,0 +1,390 @@
+System setup
+============
+
+To use ewatercycle package you need to setup the system with software
```
Comment (SarahAlidoost) [inline docs/system_setup.rst]:
```suggestion
To use eWaterCycle package you need to setup the system with software
```

File: docs/system_setup.rst [hunk 2]
```diff
@@ -0,0 +1,390 @@
...
+and data.
+
+This chapter is for system administators or Research Software Engineers who need to setup a system for the eWatercycle platform.
```
Comment (SarahAlidoost) [inline docs/system_setup.rst]:
```suggestion
This chapter is for system administrators or Research Software Engineers who need to set up a system for the eWaterCycle platform.
```

File: docs/system_setup.rst [hunk 3]
```diff
@@ -0,0 +1,390 @@
...
+7.  `Model container images <#model-container-images>`__
+8.  `Download example parameter sets <#download-example-parameter-sets>`__
+9.  `Prepare other parameter sets <#prepare-other-parameter-sets>`_
+10. `Download example forcing <#download-example-forcing>`__
+11. `Download observation data <#download-observation-data>`__
+
+Conda environment
+-----------------
+
+The ewatercycle Python package uses a lot of geospatial dependencies
```
Comment (SarahAlidoost) [inline docs/system_setup.rst]:
```suggestion
The eWaterCycle Python package uses a lot of geospatial dependencies
```

File: docs/system_setup.rst [hunk 4]
```diff
@@ -0,0 +1,390 @@
...
+which can be installed easiest using `Conda <https://conda.io/>`__ package
```
Comment (SarahAlidoost) [inline docs/system_setup.rst]:
```suggestion
which can be installed using `Conda <https://conda.io/>`__ package
```

File: docs/system_setup.rst [hunk 5]
```diff
@@ -0,0 +1,390 @@
...
+management system.
+
+Instal Conda by using the `miniconda
+installer <https://docs.conda.io/en/latest/miniconda.html>`__. For more thourough instuctions see `HPC to cluster <hpc_to_cluster.rst>`_ chapter.
```
Comment (SarahAlidoost) [inline docs/system_setup.rst]: I think that the instructions in HPC to cluster should be removed from this part. They are not about conda installation.

File: docs/system_setup.rst [hunk 6]
```diff
@@ -0,0 +1,390 @@
...
+
+.. code:: shell
+
+    conda env create --file environment.yml
+
+.. code:: shell
+
+    conda activate ewatercycle
+
+Do not forget that any terminal or Jupyter kernel should activate the conda environment before the ewatercycle Python package can be used.
```
Comment (SarahAlidoost) [inline docs/system_setup.rst]:
```suggestion
Do not forget that any terminal or Jupyter kernel should activate the conda environment before the eWaterCycle Python package can be used.
```

File: docs/system_setup.rst [hunk 7]
```diff
@@ -0,0 +1,390 @@
...
+
+Install ewatercycle package
```
Comment (SarahAlidoost) [inline docs/system_setup.rst]:
```suggestion
Install eWaterCycle package
```

File: docs/system_setup.rst [hunk 8]
```diff
@@ -0,0 +1,390 @@
...
+    # TODO once released install from PyPi instead of GitHub
+    # pip install ewatercycle
+    pip install git+https://github.com/eWaterCycle/ewatercycle.git#egg=ewatercycle
+
+
+Configure ESMValTool
+--------------------
+
+ESMValTool is used to generate forcing (temperature, precipitation,
+etc.) files from climate data sets for hydrological models. The
```
Comment (SarahAlidoost) [inline docs/system_setup.rst]:
```suggestion
etc.) files from climate data for hydrological models. The
```

File: docs/system_setup.rst [hunk 9]
```diff
@@ -0,0 +1,390 @@
...
+ESMValTool has been installed as a dependency of the package.
+
+See https://docs.esmvaltool.org/en/latest/quickstart/configuration.html
+how configure ESMValTool
```
Comment (SarahAlidoost) [inline docs/system_setup.rst]:
```suggestion
how configure ESMValTool.
```

File: docs/system_setup.rst [hunk 10]
```diff
@@ -0,0 +1,390 @@
...
+
+Download climate data
+---------------------
+
+The ERA5 and ERA-Interim climate datasets can be used to generate
+forcings
```
Comment (SarahAlidoost) [inline docs/system_setup.rst]:
```suggestion
The ERA5 and ERA-Interim data can be used to generate
forcings.
```

File: docs/system_setup.rst [hunk 11]
```diff
@@ -0,0 +1,390 @@
...
+To download ERA5 data files you can use the
+`era5cli <https://era5cli.readthedocs.io/>`__ tool.
+
+.. code:: shell
+
+    pip install era5cli
+
+To run the example notebooks the hourly ERA5 data files for year 1990
+and 1991 and for variables: pr, psl, tas, taxmin, tasmax, tdps, uas,
+vas, rsds, rsdt and fx orog need to be downloaded with
```
Comment (SarahAlidoost) [inline docs/system_setup.rst]:
```suggestion
As an example, the hourly ERA5 data for the years 1990
and 1991 and for variables pr, psl, tas, taxmin, tasmax, tdps, uas,
vas, rsds, rsdt and fx orog are downloaded as:
```
```suggestion
To run the example notebooks the hourly ERA5 data files for year 1990
and 1991 and for variables: pr, psl, tas, taxmin, tasmax, tdps, uas,
vas, rsds, rsdt and fx orog need to be downloaded with
```

File: docs/system_setup.rst [hunk 12]
```diff
@@ -0,0 +1,390 @@
...
+
+ERA-Interim
+~~~~~~~~~~~
+
+ERA-Interim has been superseeded by ERA5, but could be useful for
+reproduction studies and its smaller size. The ERA-Interim data files
+can be downloaded at
+https://www.ecmwf.int/en/forecasts/datasets/reanalysis-datasets/era-interim
+
+Installing a container engine requires root permission on the machine.
```
Comment (SarahAlidoost) [inline docs/system_setup.rst]: This line should be moved to the next section.

File: docs/system_setup.rst [hunk 13]
```diff
@@ -0,0 +1,390 @@
...
+
+    # TODO is era5cli command is correct?
+    cd <ESMValTool raw directory>
+    era5cli hourly --startyear 1990 --endyear 1991 --variables pr psl tas taxmin tasmax tdps uas vas rsds rsdt orog
+    cd -
+
+.. code:: shell
+
+    # TODO cmorize files?
+
```
Comment (SarahAlidoost) [inline docs/system_setup.rst:114]: Hourly ERA5 does not need cmorizer anymore. But we need to generate daily cmorized data: using recipe_era5.yml or recipe_daily_era5.yml if this [PR] in esmvaltool (https://github.com/ESMValGroup/ESMValTool/pull/2182) gets merged.

File: docs/system_setup.rst [hunk 14]
```diff
@@ -0,0 +1,390 @@
...
+ERA-Interim
+~~~~~~~~~~~
+
+ERA-Interim has been superseeded by ERA5, but could be useful for
+reproduction studies and its smaller size. The ERA-Interim data files
+can be downloaded at
+https://www.ecmwf.int/en/forecasts/datasets/reanalysis-datasets/era-interim
```
Comment (SarahAlidoost) [inline docs/system_setup.rst:121]: We can download the ERA-Interim using [download script](https://github.com/ESMValGroup/ESMValTool/blob/main/esmvaltool/cmorizers/obs/download_scripts/download_era_interim.py) in esmvaltool and then cmorize it by [cmorizer](https://docs.esmvaltool.org/en/latest/input.html#using-a-cmorizer-script).

File: docs/system_setup.rst [hunk 15]
```diff
@@ -0,0 +1,390 @@
...
+11. `Download observation data <#download-observation-data>`__
+
+Conda environment
+-----------------
+
+The ewatercycle Python package uses a lot of geospatial dependencies
+which can be installed easiest using `Conda <https://conda.io/>`__ package
+management system.
+
+Instal Conda by using the `miniconda
```
Comment (SarahAlidoost) [inline docs/system_setup.rst]:
```suggestion
Install Conda by using the `miniconda
```

File: docs/system_setup.rst [hunk 16]
```diff
@@ -0,0 +1,390 @@
...
+reproduction studies and its smaller size. The ERA-Interim data files
+can be downloaded at
+https://www.ecmwf.int/en/forecasts/datasets/reanalysis-datasets/era-interim
+
+Installing a container engine requires root permission on the machine.
+
+Install container engine
+------------------------
+
+In ewatercycle package the hydrological models are run in containers
```
Comment (SarahAlidoost) [inline docs/system_setup.rst]:
```suggestion
In eWaterCycle package, the hydrological models are run in containers
```

File: docs/system_setup.rst [hunk 17]
```diff
@@ -0,0 +1,390 @@
...
+
+Docker
+~~~~~~
+
+Install Docker using
+`instructions <https://docs.docker.com/engine/install/>`__. Docker
+should be configured so it can be `called without
+sudo <https://docs.docker.com/engine/install/linux-postinstall/#manage-docker-as-a-non-root-user>`__
+
+Configure ewatercycle
```
Comment (SarahAlidoost) [inline docs/system_setup.rst]:
```suggestion
Configure eWaterCycle
```

File: docs/system_setup.rst [hunk 18]
```diff
@@ -0,0 +1,390 @@
...
+---------------------
+
+The ewatercycle package simplifies the API by reading some of the
+directories and other configurations from a configuration file.
```
Comment (SarahAlidoost) [inline docs/system_setup.rst]:
```suggestion
The eWaterCycle package simplifies the API by reading some of the
directories and settings from a configuration file.
```

File: docs/system_setup.rst [hunk 19]
```diff
@@ -0,0 +1,390 @@
...
+it should be copied to ``~/.config/ewatercycle/ewatercycle.yaml`` .
+
+To make the ewatercycle configuration available to all users on the
+system it should be copied to ``/etc/ewatercycle.yaml`` .
+
+Configuration file for Cartesius system
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Users part of the ewatercycle project on the Cartesius system of
+SURFSara can use the following configuration file:
```
Comment (SarahAlidoost) [inline docs/system_setup.rst]:
```suggestion
Users part of the eWaterCycle project can use the following configurations on the Cartesius system of
SURFSara:
```

File: docs/system_setup.rst [hunk 20]
```diff
@@ -0,0 +1,390 @@
...
+   singularity_dir: /projects/0/wtrcycle/singularity-images
+   output_dir: /scratch/shared/ewatercycle
+   grdc_location: /projects/0/wtrcycle/GRDC
+   parameterset_dir: /projects/0/wtrcycle/parameter-sets
+
+Configuration file for ewatecycle Jupyter machine
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+On systems constructed with ewatercycle application on SURF Research
+Cloud can use the following configuration file:
```
Comment (SarahAlidoost) [inline docs/system_setup.rst]:
```suggestion
Users can use the following configurations on systems constructed with eWaterCycle application on SURF Research
Cloud:
```

File: docs/system_setup.rst [hunk 21]
```diff
@@ -0,0 +1,390 @@
...
+   container_engine: singularity
+   singularity_dir: /mnt/data/singularity-images
+   output_dir: /scratch
+   grdc_location: /mnt/data/GRDC
+   parameterset_dir: /mnt/data/parameter-sets
+
+Model container images
+----------------------
+
+As hydrological models run in container their container images should be
```
Comment (SarahAlidoost) [inline docs/system_setup.rst]:
```suggestion
As hydrological models run in containers, their container images should be
```

File: docs/system_setup.rst [hunk 22]
```diff
@@ -0,0 +1,390 @@
...
+Singularity images should be stored in configured directory
+(``ewatercycle.CFG['singularity_dir']``) and can build from Docker with:
+
+.. code:: ipython3
+
+    !cd {ewatercycle.CFG['singularity_dir']}
+    !singularity build ewatercycle-lisflood-grpc4bmi.sif docker://ewatercycle/lisflood-grpc4bmi:latest
+    !singularity build ewatercycle-pcrg-grpc4bmi.sif docker://ewatercycle/pcrg-grpc4bmi:latest
+    !singularity build ewatercycle-marrmot-grpc4bmi.sif docker://ewatercycle/marrmot-grpc4bmi:latest
+    !singularity build ewatercycle-wflow-grpc4bmi.sif docker://ewatercycle/wflow-grpc4bmi:latest
```
Comment (SarahAlidoost) [inline docs/system_setup.rst]: Shall we use the model versions as the tag instead of `latest`?

File: docs/system_setup.rst [hunk 23]
```diff
@@ -0,0 +1,390 @@
...
+The ``parameter_set`` variable can be passed to a model class
+constructor.
+
+Prepare other parameter sets
+----------------------------
+
+The example pararmeter sets downloaded in previous section are nice to show off the plaform features, but are a bit small.
+To preform more advanced experiments additonal parameter sets are needed.
+Users could use :py:class:`ewatercycle.parameter_sets.ParameterSet` to construct parameter sets themselves.
+Or they can be made available via :py:func:`ewatercycle.parameter_sets.available_parameter_sets` and :py:func:`ewatercycle.parameter_sets.get_parameter_set` by extending the configuration file (ewatercycle.yaml).
```
Comment (SarahAlidoost) [inline docs/system_setup.rst]:
```suggestion
The example parameter sets downloaded in the previous section are nice to show off the platform features but are a bit small.
To perform more advanced experiments, additional parameter sets are needed.
Users could use :py:class:`ewatercycle.parameter_sets.ParameterSet` to construct parameter sets themselves.
Or they can be made available via :py:func:`ewatercycle.parameter_sets.available_parameter_sets` and :py:func:`ewatercycle.parameter_sets.get_parameter_set` by extending the configuration file (ewatercycle.yaml).
```

File: docs/system_setup.rst [hunk 24]
```diff
@@ -0,0 +1,390 @@
...
+
+A new parameter set should be added as a key/value pair in the `parameter_sets` map of the configuration file.
+The key should be a unique string on the current system.
+The value is a dictionary with the following items:
+
+* directory: Location on disk where files of parameter set are stored. If Path is relative then relative to :py:const:`ewatercycle.CFG['parameterset_dir']`.
+* config: Model configuration file which uses files from directory. If Path is relative then relative to :py:const:`ewatercycle.CFG['parameterset_dir']`.
+* doi: Persistent identifier of parameter set. For a example a DOI for a Zenodo record.
+* target_model: Name of model that parameter set can work with
+* supported_model_versions: Set of model versions that are supported by this parameter set. If not set then parameter set will be supported by all versions of model
```
Comment (SarahAlidoost) [inline docs/system_setup.rst]:
```suggestion
* directory: Location on disk where files of the parameter set are stored. If Path is relative then relative to :py:const:`ewatercycle.CFG['parameterset_dir']`.
* config: Model configuration file which uses files from directory. If Path is relative then relative to :py:const:`ewatercycle.CFG['parameterset_dir']`.
* doi: Persistent identifier of the parameter set. For example a DOI for a Zenodo record.
* target_model: Name of the model that parameter set can work with
* supported_model_versions: Set of model versions that are supported by this parameter set. If not set then parameter set will be supported by all versions of model
```

**Review by SarahAlidoost — CHANGES_REQUESTED** [review]
@sverhoeven many thanks. I liked that you replaced the notebook with the rst file. just some comments on the text.

File: docs/system_setup.rst (hunk 18 above)
Comment (Peter9192) [reply docs/system_setup.rst]: Maybe elaborate a bit more on "the API" here

**Comment by SarahAlidoost** [comment]
One more comment about era5cli commandline: the variable names should be ERA5 original name and not cmor short names, see an example of download script [here](https://github.com/eWaterCycle/era5-comparison/pull/23/files#diff-c06b5270aba86875d58d636853ff432f2964c4d8d296f96530cfddac39c4be0b). Also, users should register at CDS. I think it is better to refer the user to era5cli [instructions](https://era5cli.readthedocs.io/en/stable/instructions.html).

File: docs/system_setup.rst (hunk 5 above)
Comment (sverhoeven) [reply docs/system_setup.rst]: There is a install conda section at https://ewatercycle.readthedocs.io/en/latest/hpc_to_cluster.html#first-run-preparations, which I tried to point to. Now updated hpc_to_cluster.rst with same install eWaterCycle package instructions

File: docs/system_setup.rst (hunk 12 above)
Comment (sverhoeven) [reply docs/system_setup.rst]: done

File: docs/system_setup.rst (hunk 22 above)
Comment (sverhoeven) [reply docs/system_setup.rst]: Now using same labels as used in model classes

File: docs/system_setup.rst (hunk 13 above)
Comment (sverhoeven) [reply docs/system_setup.rst:114]: Did it

File: docs/system_setup.rst (hunk 14 above)
Comment (sverhoeven) [reply docs/system_setup.rst:124]: Spelled it out

Comment (sverhoeven) [reply docs/system_setup.rst:121]: Thanks pointed to those

File: docs/system_setup.rst (hunk 11 above)
Comment (sverhoeven) [reply docs/system_setup.rst]: added

**Review by SarahAlidoost — APPROVED** [review]
Looks good to me :)

**Pull request merged**